File: src/utils/etfHoldingPreset.ts
/**
 * ETF_HOLDING_ALGORITHM (1..10, float) is the single knob driving ETF
 * holdings surfacing aggressiveness end-to-end. Structurally identical to
 * algorithmPreset.ts: parse env var → clamp → interpolate against the table
 * in models/etf_holding_presets.json. Loaded once per process via this
 * module so all three call sites (etfDiscovery.ts, etfHoldings.ts,
 * holdings/route.ts) resolve the same preset.
 */
import { readFileSync } from 'fs';
import { join } from 'path';
import { createLogger } from '@/utils/logger';

const logger = createLogger('utils/etfHoldingPreset');

export interface EtfHoldingPreset {
    /** Minimum ETF GPS score for an ETF to qualify as "hot" — replaces
     *  the hardcoded ETF_GPS_THRESHOLD constant in etfDiscovery.ts. */
    etfGpsThreshold: number;
    /** Top-N qualifying ETFs to keep after sorting by score — replaces
     *  the hardcoded .slice(0, 10) in etfDiscovery.ts. */
    topNEtfs: number;
    /** Per-ETF holdings fetch cap AND post-dedup scoring cap — replaces
     *  the ETF_HOLDING_MAX_TICKERS env var. */
    maxTickers: number;
    /** Minimum holding GPS score to mark as surfaced — replaces the
     *  ETF_HOLDING_GPS_SURFACE_VALUE env var. */
    gpsSurfaceValue: number;
    /** Minimum predicted-change % for a holding to surface — replaces
     *  the ETF_HOLDING_MIN_PRED_CHANGE env var. */
    minPredChangePct: number;
}

export interface ResolvedEtfHoldingAlgorithm extends EtfHoldingPreset {
    level: number;
}

const DEFAULT_LEVEL = 5;

let cachedTable: Record<string, EtfHoldingPreset> | null = null;

function loadTable(): Record<string, EtfHoldingPreset> {
    if (cachedTable) return cachedTable;
    const path = join(process.cwd(), 'models', 'etf_holding_presets.json');
    const raw = readFileSync(path, 'utf8');
    cachedTable = JSON.parse(raw) as Record<string, EtfHoldingPreset>;
    return cachedTable;
}

/** Parse ETF_HOLDING_ALGORITHM as a float; clamp to [1,10]; default to 5.
 *  Fractional levels are accepted and interpolated by loadEtfHoldingPreset. */
export function clampEtfHoldingLevel(raw: string | undefined): number {
    if (raw === undefined || raw === null || raw === '') return DEFAULT_LEVEL;
    const n = parseFloat(raw);
    if (Number.isNaN(n) || n < 1 || n > 10) {
        logger.warn(`Invalid ETF_HOLDING_ALGORITHM='${raw}', defaulting to ${DEFAULT_LEVEL}`);
        return DEFAULT_LEVEL;
    }
    return n;
}

function lerp(a: number, b: number, t: number): number {
    return a + (b - a) * t;
}

/** Return the preset for the given level. Integer levels read the table row
 *  directly; fractional levels are linearly interpolated between the two
 *  adjacent integer rows. topNEtfs and maxTickers are rounded to integers
 *  since they're counts; the threshold fields stay as floats. */
export function loadEtfHoldingPreset(level: number): EtfHoldingPreset {
    const table = loadTable();
    const lowerKey = Math.floor(level);
    const upperKey = Math.ceil(level);
    const lowerEntry = table[String(lowerKey)] ?? table[String(DEFAULT_LEVEL)];
    if (lowerKey === upperKey) return lowerEntry;
    const upperEntry = table[String(upperKey)] ?? lowerEntry;
    const t = level - lowerKey;
    return {
        etfGpsThreshold:  lerp(lowerEntry.etfGpsThreshold,  upperEntry.etfGpsThreshold,  t),
        topNEtfs:         Math.round(lerp(lowerEntry.topNEtfs,   upperEntry.topNEtfs,    t)),
        maxTickers:       Math.round(lerp(lowerEntry.maxTickers, upperEntry.maxTickers,  t)),
        gpsSurfaceValue:  lerp(lowerEntry.gpsSurfaceValue,  upperEntry.gpsSurfaceValue,  t),
        minPredChangePct: lerp(lowerEntry.minPredChangePct, upperEntry.minPredChangePct, t),
    };
}

/** One-shot: resolve env var → level + preset, suitable for passing into
 *  sharedContext or echoing into a meta block. */
export function resolveEtfHoldingAlgorithm(raw: string | undefined): ResolvedEtfHoldingAlgorithm {
    const level = clampEtfHoldingLevel(raw);
    return { level, ...loadEtfHoldingPreset(level) };
}

File: scripts/etf_holding_preset.py
"""Python mirror of src/utils/etfHoldingPreset.ts.

Loads models/etf_holding_presets.json and resolves a level (1.0..10.0,
fractional allowed) to a preset dict via the same linear-interpolation
rule as the TS side. Used by scripts/update_predictions.py so the per-user
ETF holding sync uses identical thresholds to the Next.js routes.
"""
from __future__ import annotations

import json
import logging
import os
from pathlib import Path

_LOG = logging.getLogger(__name__)

_PROJECT_ROOT = Path(__file__).resolve().parent.parent
_PRESETS_PATH = _PROJECT_ROOT / "models" / "etf_holding_presets.json"
_DEFAULT_LEVEL = 5.0

_cached_table: dict | None = None


def _load_table() -> dict:
    global _cached_table
    if _cached_table is None:
        with _PRESETS_PATH.open() as f:
            _cached_table = json.load(f)
    return _cached_table


def clamp_etf_holding_level(raw: str | None) -> float:
    """Parse ETF_HOLDING_ALGORITHM; clamp to [1,10]; default to 5 on bad input."""
    if not raw:
        return _DEFAULT_LEVEL
    try:
        n = float(raw)
    except ValueError:
        _LOG.warning("Invalid ETF_HOLDING_ALGORITHM=%r, defaulting to %s", raw, _DEFAULT_LEVEL)
        return _DEFAULT_LEVEL
    if n < 1 or n > 10:
        _LOG.warning("ETF_HOLDING_ALGORITHM=%r out of range [1,10], defaulting to %s", raw, _DEFAULT_LEVEL)
        return _DEFAULT_LEVEL
    return n


def _lerp(a: float, b: float, t: float) -> float:
    return a + (b - a) * t


def load_etf_holding_preset(level: float) -> dict:
    """Return the preset for the given level; interpolate for fractional levels."""
    table = _load_table()
    lo, hi = int(level // 1), int(-(-level // 1))  # floor / ceil
    if lo < 1: lo = 1
    if hi > 10: hi = 10
    lower = table[str(lo)]
    if lo == hi:
        return dict(lower)
    upper = table[str(hi)]
    t = level - lo
    return {
        "etfGpsThreshold":  _lerp(lower["etfGpsThreshold"],  upper["etfGpsThreshold"],  t),
        "topNEtfs":         round(_lerp(lower["topNEtfs"],   upper["topNEtfs"],   t)),
        "maxTickers":       round(_lerp(lower["maxTickers"], upper["maxTickers"], t)),
        "gpsSurfaceValue":  _lerp(lower["gpsSurfaceValue"],  upper["gpsSurfaceValue"],  t),
        "minPredChangePct": _lerp(lower["minPredChangePct"], upper["minPredChangePct"], t),
    }


def resolve_etf_holding_algorithm(raw: str | None = None) -> dict:
    """One-shot: read env (default ETF_HOLDING_ALGORITHM), clamp, interpolate."""
    if raw is None:
        raw = os.getenv("ETF_HOLDING_ALGORITHM")
    level = clamp_etf_holding_level(raw)
    preset = load_etf_holding_preset(level)
    preset["level"] = level
    return preset
